pkg/ccl/backupccl/testdata/backup-restore/procedures

```diff
@@ -0,0 +1,393 @@
+BACKUP INTO 'nodelocal://1/test/'
+----
+
+query-sql
+WITH descs AS (
+  SHOW BACKUP LATEST IN 'nodelocal://1/test/'
+)
+SELECT
+  database_name, parent_schema_name, object_name, object_type, is_full_cluster
+FROM
+  descs
+WHERE
+  database_name = 'db1'
+
+----
+db1 <nil> public schema true
+db1 <nil> sc1 schema true
+db1 sc1 tbl1 table true
+db1 sc1 enum1 type true
+db1 sc1 _enum1 type true
+db1 sc1 sq1 table true
+db1 sc1 p1 function true
+db1 <nil> sc2 schema true
+db1 sc2 tbl2 table true
+db1 sc2 p2 function true
+
+query-sql
+SELECT create_statement FROM [SHOW CREATE PROCEDURE sc1.p1]
+----
+CREATE PROCEDURE sc1.p1(IN a db1.sc1.enum1)
+	LANGUAGE SQL
+	AS $$
+	SELECT a FROM db1.sc1.tbl1;
+	SELECT nextval('sc1.sq1'::REGCLASS);
+$$
+
+query-sql
+CALL sc1.p1('Good'::sc1.enum1)
+----
+
+query-sql
+SELECT currval('sc1.sq1')
+----
+1
+
+# Start a new cluster with the same IO dir.
+new-cluster name=s2 share-io-dir=s1
+----
+
+# Restore into the new cluster.
+exec-sql cluster=s2
+RESTORE FROM LATEST IN 'nodelocal://1/test/'
+----
+
+exec-sql
+USE db1
+----
+
+# Make sure procedure ids in signature and body are rewritten.
+# 1. argument type id is rewritten so that type name is deserialized correctly.
+# 2. db name in qualified name is rewritten.
+# 3. sequence id is rewritten so that sequence name is deserialized correctly.
+query-sql
+SELECT create_statement FROM [SHOW CREATE PROCEDURE sc1.p1]
+----
+CREATE PROCEDURE sc1.p1(IN a db1.sc1.enum1)
+	LANGUAGE SQL
+	AS $$
+	SELECT a FROM db1.sc1.tbl1;
+	SELECT nextval('sc1.sq1'::REGCLASS);
+$$
+
+# Make sure procedure signature is rewritten in schema descriptor so that
+# procedure can be resolved and executed.
+exec-sql
+CALL sc2.p2(456)
+----
+
+query-sql
+SELECT * FROM sc2.tbl2
+----
+123
+456
+
+# Make sure dependency IDs are rewritten.
+# Note that technically this only tests forward-reference IDs in depended-on
+# objects are rewritten. But since we have cross-references validation, so this
+# also means back-references in UDF descriptor are good.
+exec-sql
+DROP SEQUENCE sc1.sq1
+----
+pq: cannot drop sequence sq1 because other objects depend on it
+
+exec-sql
+DROP TABLE sc1.tbl1
+----
+pq: cannot drop table tbl1 because other objects depend on it
+
+exec-sql
+ALTER TABLE sc1.tbl1 RENAME TO tbl1_new
+----
+pq: cannot rename relation "sc1.tbl1" because function "p1" depends on it
+HINT: consider dropping "p1" first.
+
+exec-sql
+ALTER TABLE sc1.tbl1 SET SCHEMA sc2;
+----
+pq: cannot set schema on relation "tbl1" because function "p1" depends on it
+HINT: consider dropping "p1" first.
+
+exec-sql
+DROP TYPE sc1.enum1
+----
+pq: cannot drop type "enum1" because other objects ([db1.sc1.p1]) still depend on it
+
+# Make sure that backup and restore individual tables from schema with procedure
+# does not crash.
+new-cluster name=s3
+----
+
+exec-sql cluster=s3
+CREATE DATABASE db1;
+CREATE SCHEMA sc1;
+CREATE TABLE sc1.t(a INT PRIMARY KEY);
+CREATE PROCEDURE sc1.p() LANGUAGE SQL AS $$ SELECT 1 $$;
```

Comment:
just to confirm, a table descriptor will never depend on a stored procedure? If not, we may need to add to a `skip_missing_procedures` option to restore. We had to add a similar [option ](https://github.com/cockroachdb/cockroach/commit/86be20183ab9816c4c59fb3d7b26dd6e0262e540)for udfs after an escalation.

Reply:
A procedure can only be invoked by a `CALL` statement, which is not allowed in a scalar expression like a DEFAULT expression or computed column.